modules/refined/src/main/scala/atto/syntax/ParserRefinedOps.scala

```diff
@@ -0,0 +1,34 @@
+package atto
+package syntax
+
+import eu.timepit.refined.api.{Refined, Validate}
+import eu.timepit.refined.refineV
+import scala.List
+import scala.language.implicitConversions
+
+trait ParserRefinedOps[T] {
+  val self: Parser[T]
+
+  def refined[P](implicit ev: Validate[T, P]): Parser[T Refined P] = {
+    import Parser.{Failure, State, Success, TResult}
+
+    self.flatMap(raw => new Parser[T Refined P] {
+      override def apply[R](st0: State, kf: Failure[R], ks: Success[T Refined P, R]): TResult[R] =
```

Comment:
I think you can use `ok` and `err` rather than parser internals here.

```scala
self.flatMap(refineV(_).fold(err, ok)) // you'll probably need some type arguments
```